cmd/hope/node/status.go

```diff
@@ -0,0 +1,200 @@
+}
+
+var statusCmd = &cobra.Command{
+	Use:   "status",
+	Short: "Checks the status of nodes from the manifest, and prints names, along with statuses.",
+	Args:  cobra.RangeArgs(0, 1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		var nodes []hope.Node
+		var err error
+		if len(args) == 1 {
+			if len(*statusCmdTypeSlice) != 0 {
+				return fmt.Errorf("cannot provide a node and a type")
+			}
+
+			nodeName := args[0]
+			node, err := utils.GetBareNode(nodeName)
+			if err != nil {
+				return err
+			}
+
+			nodes = []hope.Node{node}
+		} else {
+			if len(*statusCmdTypeSlice) == 0 {
+				*statusCmdTypeSlice = []string{
+					hope.NodeRoleHypervisor.String(),
+					hope.NodeRoleLoadBalancer.String(),
+					hope.NodeRoleMaster.String(),
+					hope.NodeRoleMasterAndNode.String(),
+					hope.NodeRoleNode.String(),
+				}
+			}
+
+			nodes, err = utils.GetBareNodeTypes(*statusCmdTypeSlice)
+			if err != nil {
+				return err
+			}
+		}
+
+		// If there are any Kubernetes nodes to be checked, cache a kubectl
+		//   instance.
+		var kubectl *kubeutil.Kubectl
+		defer func() {
+			if kubectl != nil {
+				kubectl.Destroy()
+			}
+		}()
+
+		for _, node := range nodes {
+			if node.IsKubernetesNode() {
+				kubectl, err = utils.KubectlFromAnyMaster()
+				if err != nil {
+					return err
+				}
+				break
+			}
+		}
+
+		nodeStatuses := map[string]hope.NodeStatus{}
+		shouldFail := false
+		for _, node := range nodes {
+			nodeStatuses[node.Name] = hope.NodeStatusUnavailable
+
+			switch node.Role {
+			case hope.NodeRoleLoadBalancer.String():
+				var status hope.NodeStatus
+				status, err = loadBalancerNodeStatus(node)
+				if err != nil {
+					return err
+				}
+
+				nodeStatuses[node.Name] = status
+			case hope.NodeRoleMaster.String(),
+					hope.NodeRoleMasterAndNode.String(),
+					hope.NodeRoleNode.String():
+				var status hope.NodeStatus
+				status, err = kubernetesNodeStatus(kubectl, node)
+				if err != nil {
+					return err
+				}
+
+				nodeStatuses[node.Name] = status
+			case hope.NodeRoleHypervisor.String():
+				var status hope.NodeStatus
+				status, err = hypervisorNodeStatus(node)
+				if err != nil {
+					return err
+				}
+
+				nodeStatuses[node.Name] = status
+			default:
+				return fmt.Errorf("unknown node type: %s", node.Role)
+			}
+
+			if nodeStatuses[node.Name] != hope.NodeStatusHealthy {
+				shouldFail = true
+			}
+		}
+
+		// Order output by the order in the yaml file, rather than iterating
+		//   over the map.
+		writer := tabwriter.NewWriter(os.Stdout, 1, 1, 1, ' ', 0)
+		fmt.Fprintln(writer, "Node\tStatus")
+		for _, node := range nodes {
+			fmt.Fprintf(writer, "%s\t%s\t\n", node.Name, nodeStatuses[node.Name])
+		}
+		writer.Flush()
+
+		if shouldFail {
+			return errors.New("error with nodes; see output for more details")
+		}
+
+		return nil
+	},
+}
+
+func kubernetesNodeStatus(kubectl *kubeutil.Kubectl, node hope.Node) (hope.NodeStatus, error) { 
```

Comment:
Since this short circuits on healthy, it could be easily adapted to pull all the nodes' Kubelet status up front, and only end up making 1 API server call, rather than one per node.